Fix test failures due to VectorUtils isFinite check

### Description

We have a wave of failing tests due to an exception in VectorUtils like this one:

```
java.lang.AssertionError: not finite: NaN from <[NaN, NaN, NaN, NaN],[NaN, NaN, NaN, NaN]>
   >         at org.apache.lucene.util.VectorUtil.dotProduct(VectorUtil.java:68)
   >         at org.apache.lucene.codecs.lucene104.Lucene104ScalarQuantizedVectorsWriter$QuantizedFloatVectorValues.<init>(Lucene104ScalarQuantizedVectorsWriter.java:702)
   >         at org.apache.lucene.codecs.lucene104.Lucene104ScalarQuantizedVectorsWriter.mergeOneFieldToIndex(Lucene104ScalarQuantizedVectorsWriter.java:434)
   >         at org.apache.lucene.codecs.lucene104.Lucene104ScalarQuantizedVectorsWriter.mergeOneFieldToIndex(Lucene104ScalarQuantizedVectorsWriter.java:409)
   >         at org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsWriter.mergeOneField(Lucene99HnswVectorsWriter.java:373)
   >         at org.apache.lucene.codecs.perfield.PerFieldKnnVectorsFormat$FieldsWriter.mergeOneField(PerFieldKnnVectorsFormat.java:128)
   >         at org.apache.lucene.codecs.KnnVectorsWriter.merge(KnnVectorsWriter.java:105)
...
```

    gradlew test --tests TestLucene104HnswScalarQuantizedVectorsFormat.testEmptyFloatVectorData -Dtests.seed=462DC933586DDAED -Dtests.multiplier=3 -Dtests.nightly=true -Dtests.locale=ksf-CM -Dtests.timezone=America/Nassau -Dtests.asserts=true -Dtests.file.encoding=UTF-8

### Version and environment details

_No response_

--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -89,6 +89,8 @@ Bug Fixes
 * GITHUB#14847: Allow Faiss vector format to index >2GB of vectors per-field per-segment by using MemorySegment APIs
   (instead of ByteBuffer) to copy bytes to native memory. (Kaival Parikh)
 
+* GITHUB#15213: Don't normalize centroids with zero samples, which caused divide-by-zero. (Mike Sokolov)
+
 Changes in Runtime Behavior
 ---------------------
 * GITHUB#14187: The query cache is now disabled by default. (Adrien Grand)

--- a/lucene/core/src/java/org/apache/lucene/codecs/lucene104/Lucene104ScalarQuantizedVectorsWriter.java
+++ b/lucene/core/src/java/org/apache/lucene/codecs/lucene104/Lucene104ScalarQuantizedVectorsWriter.java
@@ -529,7 +529,9 @@ public class Lucene104ScalarQuantizedVectorsWriter extends FlatVectorsWriter {
         mergedCentroid[j] += centroid[j] * vectorCount;
       }
     }
-    if (recalculate) {
+    if (totalVectorCount == 0) {
+      return 0;
+    } else if (recalculate) {
       return calculateCentroid(mergeState, fieldInfo, mergedCentroid);
     } else {
       for (int j = 0; j < mergedCentroid.length; j++) {

--- a/lucene/core/src/java/org/apache/lucene/util/VectorUtil.java
+++ b/lucene/core/src/java/org/apache/lucene/util/VectorUtil.java
@@ -65,7 +65,14 @@ public final class VectorUtil {
       throw new IllegalArgumentException("vector dimensions differ: " + a.length + "!=" + b.length);
     }
     float r = IMPL.dotProduct(a, b);
-    assert Float.isFinite(r);
+    assert Float.isFinite(r)
+        : "not finite: "
+            + r
+            + " from <"
+            + java.util.Arrays.toString(a)
+            + ","
+            + java.util.Arrays.toString(b)
+            + ">";
     return r;
   }
 

